protocol/requests/create_account.go

```diff
@@ -65,12 +78,14 @@ type WalletSecretsConfig struct {
 }
 
 func (c *CreateAccount) Validate() error {
-	return ValidateAccountCreationRequest(*c)
+	return ValidateAccountCreationRequest(*c, false)
 }
 
-func ValidateAccountCreationRequest(c CreateAccount) error {
+func ValidateAccountCreationRequest(c CreateAccount, allowEmptyDisplayName bool) error {
 	// TODO(cammellos): Add proper validation for password/displayname/etc
-	if len(c.DisplayName) == 0 {
+
+	// Empty display name is allowed during account restore
+	if len(c.DisplayName) == 0 && !allowEmptyDisplayName {
 		return ErrCreateAccountInvalidDisplayName
 	}
```

Comment:
This is a straightforward solution for negating the only one thing, but this solution also changes function signature in places where no change to the code is needed and this change is not easily extensible.

What do you think about the below suggestion?

```suggestion
	// Empty display name is allowed during account restore
	if len(c.DisplayName) == 0 && c.validate.DisplayName {
		return ErrCreateAccountInvalidDisplayName
	}
```

Add the following struct to `CreateAccount`

```go
type CreateAccountSkipValidation struct {
  DisplayName bool `json:"displayName"`
}
```

This way you can skip other validation as needed and no function signature needs to be changed.

Reply:
One thing, I think it's valuable that any struct in `protocol/requests` represents our API and it's something that the client wants to set.
In this case, as far as I can see, whether we should allow a blank display name is decided at the endpoint level (Create -> no, Restore -> yes), so we should not be adding to the request struct in this case, as it's helpful to look at those structs to understand exactly what the API is (it's not clear from the code suggestion above, maybe it's already the case though)

Reply:
Yes I agree, I'm sorry I should have been more clear, in my example I use `c.validate.DisplayName` and not `c.Validate.DisplayName` just so the values can't be manipulated outside the package. But let's be super careful, and explicitly tell the `json` en/decoder to ignore that field. Like below.

 ```go
type CreateAccount struct {
  // All the other fields

  validate CreateAccountSkipValidation`json:"-"`
}

type CreateAccountSkipValidation struct {
  DisplayName bool `json:"displayName"`
}
 ```

So the obvious downside to this approach is that you create a little more code overhead in the form of a dedicated "SkipValidation" struct, and the behaviour of the Validation function is dependent on the statefulness of the `CreateAccount` struct.

The upside is that `Validate()`'s signature does not change, and adding additional skips doesn't require additional function parameters. Validation becomes internally configurable.

Reply:
not a strong objection, but myself I would rather keep it as a separate struct, passed as a second argument, I like the idea to keep `requests` struct to be purely a representation of the API, albeit this is a small concession and looks ok

Reply:
Yeah, that sounds reasonable. Now we just need @igor-sirotin to tell us he hates all of this. 😆

Reply:
Well... I agree that my solution was not good, I just didn't want to think forward too much just because of a single edge case so far. But I'm open to make it better, so let's try 🙂

Reply:
There're 2 main complains here:
1. We don't want to add very special `allowEmptyDisplayName` argument to `Validate` function.
It's not an extendable solution and looks nasty.
2. We want to keep `requests` a representation of API.

# Option 1: `validate` struct as a field of `CreateAccount`

```go
type CreateAccount struct {
 // All the other fields
 validate CreateAccountSkipValidation`json:"-"`
}

type CreateAccountSkipValidation struct {
 DisplayName bool `json:"displayName"`
}

func (c *CreateAccount) Validate() {
	if len(c.DisplayName) == 0 && c.validate.DisplayName {
		return ErrCreateAccountInvalidDisplayName
	}
}
```

## Pros:
1. No extra arguments
2. Easy to add more validation options

## Cons:
1. Mixing API request struct with internal arguments

# Option 2: `validate` struct as argument of `Validate()`

```go
type CreateAccountSkipValidation struct {
 DisplayName bool `json:"displayName"`
}

func (c *CreateAccount) Validate(validate *CreateAccountSkipValidation) {
	if len(c.DisplayName) == 0 && validate.DisplayName {
		return ErrCreateAccountInvalidDisplayName
	}
}
```

### Pros:
1. No mixing API request struct with internal arguments
2. Easy to add more validation options

### Cons:
1. From API perspective, it's not obvious that when using a `RestoreAccount`, the `displayName` can be empty.

Also, maybe `ValidationOptions` would be a more Go-style solution? Same pros/cons, just not a struct, but list of options.

# Option 3: untie `CreateAccount` and `RestoreAccount`

If `RestoreAccount` struct wouldn't reuse `CreateAccount`, then we wouldn't have such problems. Each struct would have it's own validation rules on `DisplayName`.

We could create smth like `CreateRestoreAccountCommon` (with a better name), but without `DisplayName` here.

```go
type CreateAccount struct {
    CreateRestoreAccountCommon
    DisplayName string
}

type RestoreAccount struct {
    CreateRestoreAccountCommon
    DisplayName *string  // <-- pointer means "null" is possible
    FetchBackup bool     // Validate can consider both FetchBackup and DisplayName
}
```

-----

Sorry for so many words. 
@Samyoul @cammellos wdyt?

Reply:
2 seems fine to me, 3 looks also ok to me, not sure how idiomatic is, but makes sense, 1 I wouldn't be too keen, but to be honest it's all good on my side

Reply:
I like 2. And I like 3 also. I'm fine for 1 to be disregarded.

I can imagine it working roughly like below:

```go
type RestoreAccount struct {
    CreateRestoreAccountCommon
    DisplayName *string  // <-- pointer means "null" is possible
    FetchBackup bool     // Validate can consider both FetchBackup and DisplayName
}
```

```go

func(c *CreateRestoreAccountCommon) Validate() error {
    // Validation rules
}

func (ra *RestoreAccount) Validate() error {
    err := ra.CreateRestoreAccountCommon.Validate()
    if err != nil {
        return err
    }

    // Validation rules
}
```

Reply:
Ok then I'll take the decision to stick with Option 2, just not to change current API.
I'll add a note there that if there're more customisations are made between create/restore requests, then consider switching to option 3. Breaking it because of a single field is not rational I think.

Thanks everyone for input 🙂